src/main/java/games/strategy/triplea/oddsCalculator/ta/OddsCalculator.java

```diff
@@ -178,25 +178,34 @@ private AggregateResults calculate(final int count) {
         OddsCalculator.getUnitListByOrderOfLoss(this.attackerOrderOfLosses, attackingUnits, gameData);
     final List<Unit> defenderOrderOfLosses =
         OddsCalculator.getUnitListByOrderOfLoss(this.defenderOrderOfLosses, defendingUnits, gameData);
-    for (int i = 0; i < count && !cancelled; i++) {
-      final CompositeChange allChanges = new CompositeChange();
-      final DummyDelegateBridge bridge1 =
-          new DummyDelegateBridge(attacker, gameData, allChanges, attackerOrderOfLosses, defenderOrderOfLosses,
-              keepOneAttackingLandUnit, retreatAfterRound, retreatAfterXUnitsLeft, retreatWhenOnlyAirLeft);
-      final GameDelegateBridge bridge = new GameDelegateBridge(bridge1);
-      final MustFightBattle battle = new MustFightBattle(location, attacker, gameData, battleTracker);
-      battle.setHeadless(true);
-      battle.isAmphibious();
-      battle.setUnits(defendingUnits, attackingUnits, bombardingUnits,
-          (amphibious ? attackingUnits : new ArrayList<>()), defender, territoryEffects);
-      bridge1.setBattle(battle);
-      battle.fight(bridge);
-      aggregateResults.addResult(new BattleResults(battle, gameData));
-      // restore the game to its original state
-      gameData.performChange(allChanges.invert());
-      battleTracker.clear();
-      battleTracker.clearBattleRecords();
-    }
+    final AggregateResults aggregateResults = IntStream.range(0, count).parallel()
+        .filter(i -> !cancelled)
+        .mapToObj(i -> {
+          final CompositeChange allChanges = new CompositeChange();
+          final BattleTracker battleTracker = new BattleTracker();
+          try {
+            final DummyDelegateBridge bridge1 =
+                new DummyDelegateBridge(attacker, gameData, allChanges, attackerOrderOfLosses, defenderOrderOfLosses,
+                    keepOneAttackingLandUnit, retreatAfterRound, retreatAfterXUnitsLeft, retreatWhenOnlyAirLeft);
+            final GameDelegateBridge bridge = new GameDelegateBridge(bridge1);
+            final MustFightBattle battle = new MustFightBattle(location, attacker, gameData, battleTracker);
+            battle.setHeadless(true);
+            battle.isAmphibious();
+            battle.setUnits(defendingUnits, attackingUnits, bombardingUnits,
+                (amphibious ? attackingUnits : new ArrayList<>()), defender, territoryEffects);
+            bridge1.setBattle(battle);
+            battle.fight(bridge);
+            return new BattleResults(battle, gameData);
+          } finally {
+            // restore the game to its original state
+            gameData.performChange(allChanges.invert());
+            battleTracker.clear();
+            battleTracker.clearBattleRecords();
+          }
```

Comment:
Not sure if this `BattleTracker` cleanup is actually neccessary.
Previously it was, because there was only a single BattleTracker instance being used, but now a new one for every iteration is being created in order to avoid concurrency issues.
Also we might want to check if we could create a GameData clone that doesn't need its changes reverted after every iteration

Reply:
> Not sure if this BattleTracker cleanup is actually neccessary.

Now that we're new'ing it up on each iteration I'd agree it's not necessary and could be dropped.

> Also we might want to check if we could create a GameData clone that doesn't need its changes reverted after every iteration

Interesting question, I give this odds to be a bit more likely as unit from the virtual territories may have been removed.  OTOH since we are on a gameData clone, you never really know :smiling_imp: